### PR TITLE
Fix/Template endpoint update for subpages

### PR DESCRIPTION
**Pull request by wavvves**
## What

This PR complements #10640 and adds support to URL guessing by WordPress when updating Cart/Checkout template endpoints.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. Enable a blocks theme (eg TT3)
2. Create a page, and a subpage
3. On the frontend check for redirection from `/subpage` to `/page/subpage` (it's on by default)
4. Go to WooCommerce > Settings > Advanced and change the Cart and Checkout endpoints to match the created pages slug, and also the full path (eg `page/subpage`)
5. Verify on the page list, the correct page is marked as Cart/Checkout
6. Check that changing the endpoint to `subpage` correctly saves it as `page/subpage`

* [ ] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## WooCommerce Visibility

Required:

* [x] WooCommerce Core
* [] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:
* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [x] This PR is assigned to a milestone.

Conditional:
* [ ] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Fixed template endpoint editing to support WP redirection

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -65931,6 +65931,7 @@
 			"version": "0.6.0",
 			"resolved": "https://github.com/woocommerce/woocommerce-blocks/files/12309320/wordpress-e2e-test-utils-playwright-0.6.0.tgz",
 			"integrity": "sha512-5lo8iRuRfUOAPMRsQteSkKyMoP1IHHu6GKye8Df+YEPz1LWXcS6YRFGI3HrONBpcaHLrJC7LVav4UdB/sBPqUw==",
+			"dev": true,
```
Comment (nielslange) [inline package-lock.json]: Was this change intended or did it accidentally slip into this PR? 🤔

File: src/BlockTemplatesController.php [hunk 2]
```diff
@@ -855,12 +855,36 @@ public function update_template_permalink( $value, $option ) {
...
-		$matching_page = get_page_by_path( $permalink );
+		$matching_page = get_page_by_path( $permalink, OBJECT, 'page' );
+
+		/**
+		 * Filters whether to attempt to guess a redirect URL for a 404 request.
+		 *
+		 * Returning a false value from the filter will disable the URL guessing
+		 * and return early without performing a redirect.
+		 *
+		 * @since 5.5.0
```
Comment (nielslange) [inline src/BlockTemplatesController.php]: Why does this PHPDoc comment says `@since 5.5.0`?

File: src/BlockTemplatesController.php [hunk 3]
```diff
@@ -855,12 +855,36 @@ public function update_template_permalink( $value, $option ) {
...
+		 *
+		 * @param bool $do_redirect_guess Whether to attempt to guess a redirect URL
+		 *                                for a 404 request. Default true.
+		 */
+		if ( ! $matching_page instanceof WP_Post && apply_filters( 'do_redirect_guess_404_permalink', true ) ) {
```
Comment (nielslange) [inline src/BlockTemplatesController.php:871]: https://developer.wordpress.org/reference/hooks/do_redirect_guess_404_permalink/ shows this function definition:
```php
apply_filters( 'do_redirect_guess_404_permalink', bool $do_redirect_guess )
```
and mentioned that the default value of `$do_redirect_guess` is `true`. Based on that, I wonder if we shouldn't change the check as follows:
```suggestion
		if ( ! $matching_page instanceof WP_Post && apply_filters( 'do_redirect_guess_404_permalink' ) ) {
```

File: src/BlockTemplatesController.php [hunk 4]
```diff
@@ -855,12 +855,36 @@ public function update_template_permalink( $value, $option ) {
...
+			// If it is a subpage and url guessing is on, then we will need to get it via postname as path will not match.
```
Comment (nielslange) [inline src/BlockTemplatesController.php]: This comment implies that the page could be detected by post name, e.g. by `Cart block` instead of `/cart/cart-block`. However, this doesn't seem to be the case. Or did I misunderstand the comment?

File: src/BlockTemplatesController.php [hunk 5]
```diff
@@ -855,12 +855,36 @@ public function update_template_permalink( $value, $option ) {
...
+					'name'      => $permalink,
+				]
+			);
+
+			$matching_page = $query->have_posts() ? $query->posts[0] : null;
+		}
 
 		if ( $matching_page && 'publish' === $matching_page->post_status ) {
 			// Existing page matches given permalink; use its ID.
 			update_option( 'woocommerce_' . $page_slug . '_page_id', $matching_page->ID );
```
Comment (nielslange) [inline src/BlockTemplatesController.php:885]:
```suggestion
		// Existing page matches given permalink; use its ID.
		if ( $matching_page && 'publish' === $matching_page->post_status ) {
			update_option( 'woocommerce_' . $page_slug . '_page_id', $matching_page->ID );
```

I know that you did not touch this section, but I noticed that other checks have the comment before the check. What are your thoughts on moving this comment out of the check so that it's consistent with the other comments?

**Review by nielslange — COMMENTED** [review]
Thanks for working on this, @wavvves.

I left a few comments below. As for the testing instructions, I wonder if we should be more explicit on what pages to create to prevent something like this from happening:
Screenshot 2023-08-28 at 16 42 35

In this example, I followed the testing instructions explicitly and created a page called `page` and a sub page called `sub page`. I then applied their full path `/page/sub-page/` both to the Cart and and the Checkout page via WP Admin » WooCommerce » Settings » Advanced » Page setup.

As for the WooCommerce Visibility, isn't this a change that will live inside WooCommerce Core? AFAIC, this feature is not placed behind a [featured flag](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md#feature-plugin-flag), isn't it?

File: src/BlockTemplatesController.php (hunk 2 above)
Comment (wavvves) [reply src/BlockTemplatesController.php]: The PHPDoc was required for the hook per linting definitions, I could disable that (we have some cases) but maintained it as I think any documentation raising awareness for WP URL guessing is important

File: package-lock.json (hunk 1 above)
Comment (wavvves) [reply package-lock.json]: Something's wrong in trunk, npm install will update the lock file. I reverted one line, but failed to notice there was another line. Reverted both now 👍🏼

File: src/BlockTemplatesController.php (hunk 2 above)
Comment (nielslange) [reply src/BlockTemplatesController.php]: I wasn't referring to the PHPDoc itself, but to this part: `@since 5.5.0`. This comment indicates that this particular section had been introduced in [WooCommerce Blocks 5.5.0](https://github.com/woocommerce/woocommerce-blocks/releases/tag/v5.5.0), while it appears that it's been introduced with WooCommerce Blocks 11.0.0.

File: src/BlockTemplatesController.php (hunk 3 above)
Comment (wavvves) [reply src/BlockTemplatesController.php:871]: We can't, the url you posted does not apply to this, we need to look at the apply_filters() [docs](https://developer.wordpress.org/reference/functions/apply_filters/) where it states the $value is a required arg. 
If we don't supply that we get a critical error.

File: src/BlockTemplatesController.php (hunk 4 above)
Comment (wavvves) [reply src/BlockTemplatesController.php]: Typo, it's the post_name property, the slug. Corrected the comment

File: src/BlockTemplatesController.php (hunk 5 above)
Comment (wavvves) [reply src/BlockTemplatesController.php:885]: It states "matches" which implies the condition's verified and makes sense to be inside. The other one describes the if/else hence it being positioned outside.

**Comment by wavvves** [comment]
> I left a few comments below. As for the testing instructions, I wonder if we should be more explicit on what pages to create to prevent something like this from happening:

Even if the same page gets assigned to both, the testing steps still apply. That bug existed before #10640. I think it should be addressed on a different PR and not block the release. The worst case scenario would be disabling the checkout page (using cart instead), but it is not critical, and easily fixed by the user.
Can you create an issue for that one?

File: src/BlockTemplatesController.php (hunk 2 above)
Comment (wavvves) [reply src/BlockTemplatesController.php]: Right, copy paste from source. Corrected 👍🏼

File: src/BlockTemplatesController.php (hunk 3 above)
Comment (nielslange) [reply src/BlockTemplatesController.php:871]: Oh, my bad. You're right, Paulo.

**Review by nielslange — APPROVED** [review]
Thanks for addressing the suggestions, @wavvves. The changes look good now.

**Comment by nielslange** [comment]
> Even if the same page gets assigned to both, the testing steps still apply. That bug existed before https://github.com/woocommerce/woocommerce-blocks/pull/10640. I think it should be addressed on a different PR and not block the release. The worst case scenario would be disabling the checkout page (using cart instead), but it is not critical, and easily fixed by the user.
Can you create an issue for that one?

I created https://github.com/woocommerce/woocommerce-blocks/issues/10756 for tracking this.
